third_party/move/move-vm/types/src/loaded_data/instance_uid.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use dashmap::{
+    mapref::entry::Entry::{Occupied, Vacant},
+    DashMap,
+};
+use derivative::Derivative;
+use std::{
+    cmp::{Eq, Ordering},
+    hash::{Hash, Hasher},
+    sync::{Arc, Weak},
+};
+
+// TODO: add tests.
+// TODO: break into files.
+// TODO: micro benchmarks.
+
+pub struct GlobalCache<'a, T: Eq + Hash> {
+    instance_to_compressed: DashMap<Arc<T>, Weak<InstanceUIDInner<'a, T>>>,
```

Comment:
"compressed" is wrong here.  intern_table?

Reply:
Also, as mentioned below: why is the value an InstanceUIDInner<...>?  Why not just a Weak<T> which is downgraded Arc<T>?  (Actually, you probably want another wrapper struct like ArcKeyByPtr to enforce handling).

Reply:
Yes- "compressed is wrong". I don't really like intern_table either. It is cache in a way that keeps track only of "alive" instances.